### PR TITLE
DAOS-3669 dfs: add proper short read and hole management for byte arrays

### PR DESCRIPTION
**Pull request by mchaarawi**
This PR utilizes the IO MAP in fetch to implement short read handling and hole management for DFS access (Byte array in the dc_array layer). Note this affect only arrays that are with oid feat type DAOS_OF_ARRAY_BYTE (which is what DFS creates files with).

- The short read handling and get_size were moved from the DFS layer into the dc_array layer to combine short read handling with hole management from the IOM at the object IO layer.
- A new param is returned in the iod of the array IO "arr_nr_read" to indicate how many records were read.
- On fetch, a get_size task is created with a prepare CB that checks the iomap highest record and determines if a short read is possible:
       - If it is (we fetch on a extent bigger than what the iomap returns as highest), then the get_size is executed with a completion callback that sets the short read if it was done, and does the hole management using the iomap after
       - otherwise, if no short read is possible,  the get_size is not scheduled and we move directly to the hole processing.
- The iomap is examined to determine whether the extents fetched are valid or holes. For holes that are before the EOF, the user buffer is memset to 0. otherwise the user buffer is not modified (for extents that are valid or fetch beyond the EOF).
- Some unit tests were added to the dfs_test to exercise this code.
- This was tested also with MPI-IO and HDF5 tests that were disabled waiting for this feature.

**_Note that when the IOMAP recx buffer is smaller than what is needed, the entire fetch for the corresponding dkey/iod is re-executed since the fetch does not support a refetch with only the iomap (no bulk data). An optimization would be to allow fetch only the iomap instead of the data too. On the other hand we could go the other way and re-execute the entire array read for all dkeys, but that might be too extreme._**

### CODE REVIEW

File: src/client/array/dc_array.c [hunk 1]
```diff
@@ -175,6 +179,10 @@ free_io_params_cb(tse_task_t *task, void *data)
 	while (io_list) {
 		struct io_params *current = io_list;
 
+		if (current->iom.iom_recxs) {
+			D_FREE(current->iom.iom_recxs);
+			current->iom.iom_recxs = NULL;
```
Comment (jolivier23) [inline src/client/array/dc_array.c:184]: D_FREE sets the pointer to NULL

File: src/client/array/dc_array.c [hunk 2]
```diff
@@ -208,30 +217,35 @@ create_handle_cb(tse_task_t *task, void *data)
...
+	array->oid.hi		= args->oid.hi;
+	array->oid.lo		= args->oid.lo;
+	array->mode		= DAOS_OO_RW;
+	array->cell_size	= args->cell_size;
+	array->chunk_size	= args->chunk_size;
+	array->daos_oh		= *args->oh;
+
+	feat = daos_obj_id2feat(args->oid);
+	if (feat & DAOS_OF_ARRAY_BYTE)
+		array->byte_array = true;
```
Comment (jolivier23) [inline src/client/array/dc_array.c:230]: so is this only an option for arrays that are bytes and not for any other record size?  What if a user has same issue with their custom record size array where they want to zero holes?

File: src/client/array/dc_array.c [hunk 3]
```diff
@@ -1064,14 +1092,291 @@ create_sgl(d_sg_list_t *user_sgl, daos_size_t cell_size,
...
+struct hole_params {
+	struct io_params	*io_list;
+	tse_task_t		*ptask;
+	daos_size_t		records_req;
+	daos_size_t		array_size;
+	daos_handle_t		oh;
+};
+
+static int
+memset_sgl(uint8_t *buf, size_t len, void *args)
```
Comment (jolivier23) [inline src/client/array/dc_array.c:1104]: Minor nit: reset_buf?   memset implies being able to specify what to set it to and sgl seems irrelevant here.

File: src/client/array/dc_array.c (hunk 2 above)
Comment (mchaarawi) [reply src/client/array/dc_array.c:230]: yes this is only intended for byte arrays / dfs files.
short read handling and hole management is usually only for files. a user working with arrays (single or multi dimensional) does not usually expect this behavior.

File: src/client/array/dc_array.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/array/dc_array.c:184]: OK, will wait for more comments from others before refreshing.

File: src/client/array/dc_array.c (hunk 3 above)
Comment (mchaarawi) [reply src/client/array/dc_array.c:1104]: OK, will wait for more comments from others before refreshing.

**Comment by mchaarawi** [comment]
just another rebase ..
